core/src/replay_stage.rs

```diff
@@ -4062,6 +4062,14 @@ impl ReplayStage {
                 }
             }
 
+            if progress.is_duplicate_confirmed(*slot).unwrap_or(false) {
+                // Already duplicate confirmed, nothing left to do
+                continue;
+            }
+
+            progress.set_duplicate_confirmed_slot(*slot);
+            assert!(*frozen_hash != Hash::default());
```

Comment:
hmm, would this existing  https://github.com/anza-xyz/agave/blob/abdfbca21e8358467d2ba46e2090c8b8516bf339/core/src/replay_stage.rs#L4084-L4088 check be sufficient if we updated `duplicate_confirmed_slots` in `confirm_forks`?

Better yet, can we just remove `duplicate_confirmed_slots` and replace it with the progress map like you have here?

Reply:
I believe the idea for splitting this up into `confirm_forks` and `mark_slot_confirmed` was so `confirm_forks` is side effect free for unit testing.

And we can't replace `duplicate_confirmed_slots` with `progress_map` unfortunately because `progress_map` only holds replayed forks, however a slot could be `duplicate_confirmed` through gossip, so we have to track it separately.

Reply:
hmm is there a rason not to have progress map also track gossip duplicate confirmed forks?

Reply:
I gave it a try here: https://github.com/AshwinSekar/solana/commit/9d14e0dad3015804cd06e9a30c30aa7d1cd72f7b

The main issue is that `ProgressMap` is a wrapper around this `HashMap`:
https://github.com/anza-xyz/agave/blob/3ca67197c5cd142e355f0ad4263e6dd63d63f224/core/src/consensus/progress_map.rs#L258-L260

We insert new entries based on the bank:
https://github.com/anza-xyz/agave/blob/3ca67197c5cd142e355f0ad4263e6dd63d63f224/core/src/replay_stage.rs#L2932-L2934

Because `ForkProgress` and it's sub struct `ForkStats` are mostly comprised of information we can only get from replay:
https://github.com/anza-xyz/agave/blob/3ca67197c5cd142e355f0ad4263e6dd63d63f224/core/src/consensus/progress_map.rs#L72-L85

When we root we use bank forks to retain:
https://github.com/anza-xyz/agave/blob/3ca67197c5cd142e355f0ad4263e6dd63d63f224/core/src/consensus/progress_map.rs#L395-L398

Reply:
If we really wanted to put it in `progress_map`, we would have to put it at the top level something like:
```
 pub struct ProgressMap { 
     progress_map: HashMap<Slot, ForkProgress>, 
     duplicate_confirmed_slots:  BTreeMap<Slot, Hash>
} 
```
But at that point we're just tracking two separate things in a struct called `ProgressMap` which isn't really a progress indicator anymore.

Reply:
I see, essentially duplicate_confirmed_slots right now tracks slots that might not be in ProgressMap at all because we haven't replayed the slot, but have detected the slot via gossip